File: Java/src/main/java/QuestionBank/Solution101_200/Solution108.java
package QuestionBank.Solution101_200;

import tool.TreeNode;

/**
 * @author windows
 */
public class Solution108 {
    public TreeNode sortedArrayToBST(int[] nums) {
        return process(nums, 0, nums.length - 1);
    }

    private TreeNode process(int[] nums, int begin, int end) {
        if (begin >= end){
            if (begin == end){
                return new TreeNode(nums[begin]);
            }
            return null;
        }
        int middle = (begin + end + 1) / 2;
        TreeNode root = new TreeNode(nums[middle]);
        root.left = process(nums, begin, middle - 1);
        root.right = process(nums, middle + 1, end);
        return root;
    }




    public static void main(String[] args) {
        Solution108 solution108 = new Solution108();
        TreeNode treeNode = solution108.sortedArrayToBST(new int[]{-10, -3, 0, 5, 9});
        System.out.println("=====");
    }
}

File: C++/src/Solution101_200/Solution108.cpp
//
// Created by windows on 2023/3/1.
//

#include "Solution108.h"
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

TreeNode *Solution108::sortedArrayToBST(vector<int> &nums) {
    return helper(nums, 0, nums.size() - 1);
}

//TreeNode *Solution108::helper(vector<int> &nums, int left,  int right) {
//    if (left > right){
//        return nullptr;
//    }
//    int mid = left + (right - left) / 2;
//    TreeNode* root = new TreeNode(nums[mid]);
//    root->left = helper(nums, left, mid - 1);
//    root->right = helper(nums, mid + 1, right);
//    return root;
//}

TreeNode *Solution108::helper(vector<int> &nums, int left,  int right) {
    if (left > right){
        return nullptr;
    }
    int mid = left + (right - left) / 2;
    TreeNode* root = new TreeNode(nums[mid]);
    root->left = helper(nums, left, mid  - 1);
    root->right = helper(nums, mid + 1, right);
    return root;
}
